### DISCUSSION TITLE
ValueError: Target size (torch.Size([1490])) must be the same as input size (torch.Size([100, 1490]))

### DISCUSSION THREAD

**Discussion by joshuajennysibbu** [pyg-team/pytorch_geometric#4739] [Q&A]
I have a few questions:
-   Is negative sampling is required for edge level prediction
-  Why model return torch.Size([100, 1490],  My batch size 32 why it return 100. How to fix this issue.

```python
 class GNN(torch.nn.Module):
              def __init__(self, hidden_channels, out_channels):
                    super().__init__()
                    self.conv1 = SAGEConv((-1, -1), hidden_channels,normalize=True)
                    self.conv2 = SAGEConv((-1, -1), out_channels,normalize=True)

           def forward(self, source_,target_, edge_index_):
        
            source=source_.to(torch.float32)
        target=target_.to(torch.float32)
        edge_index_=edge_index_.to(torch.long)
        
        edge_index_=edge_index_.T
        
        print(self.conv1)
        source_conv = self.conv1((source,target),edge_index_)
        source_conv = self.conv1((source,target),edge_index_)
        print(source_conv.shape)
        source_conv = self.conv2(source_conv,edge_index_)
        
        return source_conv
`````

`loss = F.binary_cross_entropy_with_logits(output, link_labels)`

**Comment by rusty1s** [comment]
* In general, negative sampling is required if you want to infer missing links, since you need to train against both existing links and non-existing links. For edge-level property prediction, you do not need to make use of negative sampling.
* This is hard to tell. I assume you are using the `LinkNeighborLoader`? In that case, the loader will sample nodes outside of your current mini-batch that are used to aggregate information inside your GNN. You need to slice your output embeddings afterwards to fit to the correct shape: `x = x[:batch_size]`.

**Reply by joshuajennysibbu** [reply]
I'm working on bipartite graphs. I'm using Dataloader. Is LinkNeighborLoader good for link prediction? what will be y label for link prediction?

**Reply by joshuajennysibbu** [reply]
> * In general, negative sampling is required if you want to infer missing links, since you need to train against both existing links and non-existing links. For edge-level property prediction, you do not need to make use of negative sampling.
> * This is hard to tell. I assume you are using the `LinkNeighborLoader`? In that case, the loader will sample nodes outside of your current mini-batch that are used to aggregate information inside your GNN. You need to slice your output embeddings afterwards to fit to the correct shape: `x = x[:batch_size]`.


I got this error: while import LinkNeighborLoader
ImportError: cannot import name 'LinkNeighborLoader' from 'torch_geometric.loader' (/home/ubuntu/anaconda3/lib/python3.9/site-packages/torch_geometric/loader/__init__.py)

` from torch_geometric.loader import LinkNeighborLoader

    loader = LinkNeighborLoader(
    data,
    num_neighbors=[30] * 2,
    batch_size=128,
           edge_label_index=data.edge_index,
     )`

**Reply by rusty1s** [reply]
`LinkNeighborLoader` is a new feature. In order to import it, you need to install from master/nightly, see [here](https://github.com/pyg-team/pytorch_geometric#nightly-and-master).

**Reply by joshuajennysibbu** [reply]
> `LinkNeighborLoader` is a new feature. In order to import it, you need to install from master/nightly, see [here](https://github.com/pyg-team/pytorch_geometric#nightly-and-master).

I'm working on bipartite graphs. what will be y label for link prediction? I try with LinkNeighborLoader but my custom in tensor format i have (source,target) and edge_index. What i have to do? 

Data object
`d_Data=Data(x=[source,target]],edge_index=edge_index_T)`

dataloader:
`loader = LinkNeighborLoader(
    d_Data,
    num_neighbors=[30] * 2,
    batch_size=128,
    edge_label_index=d_Data.edge_index)`

I got error:
`ValueError: too many values to unpack (expected 2)`

For example:
node=[-1,0,1]   #INDEX : 0,1,2
edge_index=[[0,1],[1,2]]

Is this the right way for the edge index? I use the node index as an edge index.

**Reply by rusty1s** [reply]
You need to make use of `HeteroData` in case you want to use the `LinkNeighborLoader` in bipartite graphs.

**Reply by joshuajennysibbu** [reply]
ok... Is it possible to build a bipartite graph in dataloader class? What will be the output label for bipartite?

For example:
node=[-1,0,1] #INDEX : 0,1,2
edge_index=[[0,1],[1,2]]

Is this the right way for the edge index? I use the node index as an edge index.

**Reply by rusty1s** [reply]
Can you clarify? For a bipartite graph, you could do:
```python
data = HeteroData()
data['user'].x = ...
data['product'].x = ...
data['user, 'product'].edge_index = ...
data['product', 'user'].edge_index = ...
```
and then pass it to the `LinkNeighborLoader`:
```python
loader = LingNeighborLoader(data, num_neighbors=[10, 10],
                            edge_label_index=('user', 'product'), ...)
```

**Reply by joshuajennysibbu** [reply]
Okay... and What will be the output label for the bipartite graph?

```
node=[-1,0,1] #INDEX : 0,1,2
edge_index=[[0,1],[1,2]]

```
Is this the right way for the edge index? I use the node index as an edge index

**Reply by rusty1s** [reply]
I am not sure I understand. What do you mean with output label? For the `HeteroData`, the `edge_index` attribute will point from the nodes in the source node type to indices in the destination node type, e.g.:
```python
data = HeteroData()
data['user'].x = torch.randn(5, 16)  # 5 nodes
data['product'].x = torch.randn(2, 16)  # 2 nodes
data['user, 'product'].edge_index = torch.tensor([[0, 1, 2, 3, 4], [0, 1, 0, 1, 0]])
```

**Reply by joshuajennysibbu** [reply]
> I am not sure I understand. What do you mean with output label? For the `HeteroData`, the `edge_index` attribute will point from the nodes in the source node type to indices in the destination node type, e.g.:
> 
> ```python
> data = HeteroData()
> data['user'].x = torch.randn(5, 16)  # 5 nodes
> data['product'].x = torch.randn(2, 16)  # 2 nodes
> data['user, 'product'].edge_index = torch.tensor([[0, 1, 2, 3, 4], [0, 1, 0, 1, 0]])
> ```

I'm new to Graph Networks. how to do link prediction. Can you give me a simple code.

**Reply by rusty1s** [reply]
Please take a look at https://github.com/pyg-team/pytorch_geometric/blob/master/examples/hetero/hetero_link_pred.py.
